gh-67230: add quoting rules to csv module

smontanaro opened the pull request:
Add two quoting styles for csv dialects.
They will help to work with certain databases in particular.

<!-- gh-issue-number: gh-67230 -->
* Issue: gh-67230
<!-- /gh-issue-number -->

Automerge-Triggered-By: GH:merwok

smontanaro committed:
bpo-23041: update proposed changes to csv module.

smontanaro committed:
Merge branch 'main' into fix_issue_23041

smontanaro committed:
Merge branch 'main' into fix_issue_23041

smontanaro committed:
Merge branch 'main' into fix_issue_23041

smontanaro committed:
Merge branch 'main' into fix_issue_23041

smontanaro committed:
Merge branch 'main' into fix_issue_23041

smontanaro committed:
Merge branch 'main' into fix_issue_23041

smontanaro committed:
Merge branch 'main' into fix_issue_23041

samwyse committed:
Update test_csv.py

Improve tests of QUOTE_STRINGS and QUOTE_NOTNULL

samwyse committed:
Update csv.rst

Improve descriptions of QUOTE_NOTNULL and QUOTE_STRINGS with writer objects. Add descriptions for reader objects.

smontanaro committed:
Merge pull request #38 from samwyse/fix_issue_23041

[3.12] Fix issue 23041 (GH-67230)

smontanaro committed:
remove trailing whitespace

smontanaro committed:
Merge branch 'python:main' into fix_issue_23041

Commit:
Merge branch 'main' into fix_issue_23041

merwok commented on Doc/library/csv.rst:
```diff
@@ -327,7 +327,7 @@ The :mod:`csv` module defines the following constants:
    Instructs :class:`writer` objects to quote all non-numeric fields.
 
-   Instructs the reader to convert all non-quoted fields to type *float*.
+   Instructs :class:`reader` to convert all non-quoted fields to type *float*.
```
Little note: these blocks talk about «writer objects» and just «reader», could be harmonized?

smontanaro replied:
Thanks, will take a look unless @samwyse beats me to it.

merwok replied:
Not a blocker for this PR note, it’s the style in existing doc.

samwyse replied:
OK, Skip, I think I beat you to it.

merwok commented on Doc/library/csv.rst:
```diff
@@ -339,6 +339,24 @@ The :mod:`csv` module defines the following constants:
+.. data:: QUOTE_NOTNULL
+
+   Instructs :class:`writer` objects to quote all fields which are not
+   ``None``.  This is similar to QUOTE_ALL, except that if a
```
```suggestion
   ``None``.  This is similar to :data:`QUOTE_ALL`, except that if a
```

merwok commented on Doc/library/csv.rst:
```diff
@@ -339,6 +339,24 @@ The :mod:`csv` module defines the following constants:
+   field value is ``None`` an empty (unquoted) string is written.
+
+   Instructs :class:`reader` to interpret an empty (unquoted) field as None, and
+   otherwise behave as QUOTE_ALL.
```
```suggestion
   otherwise behave as :data:`QUOTE_ALL`.
```

Should it be behaves or behave?
not sure if the structure is _instructs ... and behaves ..._ or _instructs reader to interpret ... and to behave ..._

merwok commented on Doc/library/csv.rst:
```diff
@@ -339,6 +339,24 @@ The :mod:`csv` module defines the following constants:
+.. data:: QUOTE_STRINGS
+
+   Instructs :class:`writer` objects to always place quotes around fields
+   which are strings.  This is similar to QUOTE_NONNUMERIC, except that if a
```
```suggestion
   which are strings.  This is similar to :data:`QUOTE_NONNUMERIC`, except that if a
```

merwok commented on Doc/library/csv.rst:
```diff
@@ -339,6 +339,24 @@ The :mod:`csv` module defines the following constants:
+   field value is ``None`` an empty (unquoted) string is written.
+
+   Instructs :class:`reader` to interpret an empty (unquoted) string as None, and
+   otherwise behave as QUOTE_NONNUMERIC.
```
```suggestion
   otherwise behave as :data:`QUOTE_NONNUMERIC`.
```

merwok commented on Misc/NEWS.d/next/Library/2021-11-07-15-31-25.bpo-23041.564i32.rst:
```diff
@@ -0,0 +1,2 @@
+Add QUOTE_STRINGS and QUOTE_NOTNULL to the suite of csv module quoting
+styles.
```
```suggestion
Add ``QUOTE_STRINGS`` and ``QUOTE_NOTNULL`` to the suite of
:mod:`csv` module quoting styles.
```

merwok replied:
Or these could use this markup to create links:
```
:data:`~csv.QUOTE_STRINGS`
```
The tilde is to hide `csv` in the output, but it’s still there to create the right link.

samwyse committed:
Update csv.rst

Harmonize references to reader and writer objects.  Change to "instructs reader to interpret ... and to otherwise behave"

samwyse committed:
Update 2021-11-07-15-31-25.bpo-23041.564i32.rst

Add styling

smontanaro committed:
Merge pull request #39 from samwyse/fix_issue_23041

Fix issue 23041

merwok commented on Misc/NEWS.d/next/Library/2021-11-07-15-31-25.bpo-23041.564i32.rst:
```diff
@@ -0,0 +1,2 @@
+Add :data:`~csv.QUOTE_STRINGS` and :data:`~csv.QUOTE_NOTNULL` to the suite of :mod:`csv` module quoting
+styles.
```
```suggestion
Add :data:`~csv.QUOTE_STRINGS` and :data:`~csv.QUOTE_NOTNULL` to the suite
of :mod:`csv` module quoting styles.
```

merwok committed:
wrap line

merwok commented on Doc/library/csv.rst:
```diff
@@ -337,7 +337,25 @@ The :mod:`csv` module defines the following constants:
+   which are strings.  This is similar to :data:`QUOTE_NONNUMERIC`, except that if a
+   field value is ``None`` an empty (unquoted) string is written.
+
+   Instructs :class:`reader`objects to interpret an empty (unquoted) string as None and
```
```suggestion
   Instructs :class:`reader` objects to interpret an empty (unquoted) string as ``None`` and
```

merwok committed:
fix markup

merwok committed:
Merge branch 'main' into fix_issue_23041

GPHemsley-RELX commented on Doc/library/csv.rst:
```diff
@@ -337,7 +337,25 @@ The :mod:`csv` module defines the following constants:
+   field value is ``None`` an empty (unquoted) string is written.
+
+   Instructs :class:`reader` objects to interpret an empty (unquoted) string as ``None`` and
+   to otherwise behave as :data:`QUOTE_NONNUMERIC`.
```
These are missing a note that they were added in 3.12.
